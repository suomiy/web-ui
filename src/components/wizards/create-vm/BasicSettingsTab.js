import React from 'react';
import PropTypes from 'prop-types';
import { get } from 'lodash';
import { FormFactory } from '../../forms/FormFactory';
import { isPositiveNumber } from '../../../utils/validation';

class BasicSettingsTab extends React.Component {
  onFormChange = (newValue, target) => {
    let validMsg;

    if (this.basicFormFields[target].validate) {
      validMsg = this.basicFormFields[target].validate(newValue);
    }
    if (this.basicFormFields[target].required && newValue.trim().length === 0) {
      validMsg = 'is required';
    }

    if (validMsg) {
      validMsg = `${this.basicFormFields[target].title} ${validMsg}`;
    }

    const basicSettings = {
      ...this.props.basicSettings,
      [target]: {
        value: newValue,
        validMsg
      }
    };

    this.props.onChange(basicSettings, this.validateWizard(basicSettings));
  };

  validateWizard = values => {
    let wizardValid = true;

    // check if all required fields are defined
    const requiredKeys = Object.keys(this.basicFormFields).filter(key => this.isFieldRequired(key, values));
    const requiredKeysInValues = Object.keys(values).filter(key => this.isFieldRequired(key, values));
    if (requiredKeys.length !== requiredKeysInValues.length) {
      wizardValid = false;
    }

    // check if all fields are valid
    for (const key in values) {
      if (
        values[key].validMsg &&
        (this.basicFormFields[key].isVisible ? this.basicFormFields[key].isVisible(values) : true)
      ) {
        wizardValid = false;
        break;
      }
    }

    return wizardValid;
  };

  isFieldRequired = (key, basicVmSettings) =>
    this.basicFormFields[key].required &&
    (this.basicFormFields[key].isVisible ? this.basicFormFields[key].isVisible(basicVmSettings) : true);

  basicFormFields = {
    name: {
      title: 'Name',
      required: true
    },
    description: {
      title: 'Description',
      type: 'textarea'
    },
    namespace: {
      id: 'namespace-dropdown',
      title: 'Namespace',
      type: 'dropdown',
      defaultValue: '--- Select Namespace ---',
      choices: this.props.namespaces,
      required: true
    },
    imageSourceType: {
      id: 'image-source-type-dropdown',
      title: 'Provision Source',
      type: 'dropdown',
      defaultValue: '--- Select Provision Source ---',
      choices: [
        {
          name: 'PXE'
        },
        {
          name: 'URL'
        },
        {
          name: 'Template'
        },
        {
          name: 'Registry'
        }
      ],
      required: true
    },
    template: {
      id: 'template-dropdown',
      title: 'Template',
      type: 'dropdown',
      defaultValue: '--- Select Template ---',
      required: true,
      isVisible: basicVmSettings => get(basicVmSettings, 'imageSourceType.value') === 'Template',
      choices: this.props.templates
    },
    registryImage: {
      title: 'Registry Image',
      required: true,
      isVisible: basicVmSettings => get(basicVmSettings, 'imageSourceType.value') === 'Registry'
    },
    imageURL: {
      title: 'URL',
      required: true,
      isVisible: basicVmSettings => get(basicVmSettings, 'imageSourceType.value') === 'URL'
    },
    operatingSystem: {
      id: 'operating-system-dropdown',
      title: 'Operating System',
      type: 'dropdown',
      defaultValue: '--- Select Operating System ---',
      choices: this.props.operatingSystems,
      required: true
    },
    flavor: {
      id: 'flavor-dropdown',
      title: 'Flavor',
      type: 'dropdown',
      defaultValue: '--- Select Flavor ---',
      choices: this.props.flavors.concat([{ name: 'Custom' }]),
      required: true
    },
    memory: {
      title: 'Memory (GB)',
      required: true,
      isVisible: basicVmSettings => get(basicVmSettings, 'flavor.value', '') === 'Custom',
      validate: currentValue => (isPositiveNumber(currentValue) ? undefined : 'must be a number')
    },
    cpu: {
      title: 'CPUs',
      required: true,
      isVisible: basicVmSettings => get(basicVmSettings, 'flavor.value', '') === 'Custom',
      validate: currentValue => (isPositiveNumber(currentValue) ? undefined : 'must be a number')
    },
    workloadProfile: {
      id: 'workload-profile-dropdown',
      title: 'Workload Profile',
      type: 'dropdown',
      defaultValue: '--- Select Workload Profile ---',
      choices: this.props.workloadProfiles,
      required: true,
      help: () =>
        this.props.workloadProfiles.map(value => (
          <p key={value.id}>
            <b>{value.name}</b>: {value.description}
          </p>
        ))
    },
    startVM: {
      title: 'Start virtual machine on creation',
      type: 'checkbox',
      noBottom: true
    },
    createTemplate: {
      title: 'Create new template from configuration',
      type: 'checkbox',
      noBottom: true
    },
    cloudInit: {
      title: 'Use cloud-init',
      type: 'checkbox'
    },
    hostname: {
      title: 'Hostname',
      isVisible: basicVmSettings => get(basicVmSettings, 'cloudInit.value', false),
      required: true
    },
    authKeys: {
      title: 'Authenticated SSH Keys',
      type: 'textarea',
      isVisible: basicVmSettings => get(basicVmSettings, 'cloudInit.value', false),
      required: true
    }
  };

  render() {
    return (
      <FormFactory
        fields={this.basicFormFields}
        fieldsValues={this.props.basicSettings}
        onFormChange={this.onFormChange}
      />
    );
  }
}

BasicSettingsTab.propTypes = {
  workloadProfiles: PropTypes.array.isRequired,
  flavors: PropTypes.array.isRequired,
  operatingSystems: PropTypes.array.isRequired,
  templates: PropTypes.array.isRequired,
  namespaces: PropTypes.array.isRequired,
  basicSettings: PropTypes.object.isRequired,
  onChange: PropTypes.func.isRequired
};

export default BasicSettingsTab;
