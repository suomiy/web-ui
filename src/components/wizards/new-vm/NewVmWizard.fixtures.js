export const storageClasses = [
  {
    name: 'NFS',
    id: 'nfs'
  },
  {
    name: 'iSCSI',
    id: 'iscsi'
  },
  {
    name: 'Glusterfs',
    id: 'glusterfs'
  },
  {
    name: 'AzureDisk',
    id: 'azuredisk'
  },
  {
    name: 'Local',
    id: 'local'
  }
];

export const storages = [
  {
    id: 'disk-one',
    name: 'disk One',
    size: '10',
    storageClass: 'nfs'
  },
  {
    id: 'disk-two',
    name: 'disk Two',
    size: '15',
    storageClass: 'glusterfs'
  },
  {
    id: 'disk-three',
    name: 'disk Three',
    size: '20',
    storageClass: 'iscsi'
  }
];

export const namespaces = [
  {
    name: 'default'
  },
  {
    name: 'myproject'
  }
];

export const flavors = [
  {
    name: 'Flavor1'
  },
  {
    name: 'Flavor2'
  }
];

export const workloadProfiles = [
  {
    id: 'profile-one',
    name: 'WorkloadProfile1',
    description: 'profile description'
  },
  {
    id: 'profile-two',
    name: 'WorkloadProfile2',
    description: 'profile description'
  }
];

export const templates = [
  {
    name: 'Template1'
  },
  {
    name: 'Template2'
  }
];
