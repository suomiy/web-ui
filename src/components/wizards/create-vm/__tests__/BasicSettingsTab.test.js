import React from 'react';
import { shallow } from 'enzyme';
import BasicSettingsTab from '../BasicSettingsTab';
import { os } from '../../new-vm/OperatingSystem';
import { workloadProfiles, flavors, templates, namespaces } from '../../new-vm/NewVmWizard.fixtures';
import { validBasicSettings } from './_utils';

const testCreateVmWizard = (basicSettings = {}, onChange = null) => (
  <BasicSettingsTab
    workloadProfiles={workloadProfiles}
    flavors={flavors}
    operatingSystems={os}
    templates={templates}
    namespaces={namespaces}
    basicSettings={basicSettings}
    onChange={onChange || jest.fn()}
  />
);

const expectMockToBeCalledWith = (fn, a, b) => {
  expect(fn.mock.calls[0][0]).toEqual(a);
  expect(fn.mock.calls[0][1]).toBe(b);
};

const testFormChange = (what, value, result, valid) => {
  const onChange = jest.fn();
  const component = shallow(testCreateVmWizard({}, onChange));
  component.instance().onFormChange(value, what);

  expectMockToBeCalledWith(onChange, result, valid);
};

describe('<CreateVmWizard />', () => {
  it('renders correctly', () => {
    const component = shallow(testCreateVmWizard());
    expect(component).toMatchSnapshot();
  });

  it('is validates incomplete form', () => {
    testFormChange(
      'name',
      'someName',
      {
        name: {
          validMsg: undefined,
          value: 'someName'
        }
      },
      false
    );
  });

  it('is valid when all required fields are filled', () => {
    const onChange = jest.fn();
    const component = shallow(testCreateVmWizard(validBasicSettings, onChange));
    component.instance().onFormChange(validBasicSettings.name.value, 'name'); // trigger validation

    expectMockToBeCalledWith(onChange, validBasicSettings, true);
  });

  it('required property is validated', () => {
    testFormChange(
      'name',
      '',
      {
        name: {
          validMsg: 'Name is required',
          value: ''
        }
      },
      false
    );
  });

  it('cpu field validation is triggered', () => {
    testFormChange(
      'cpu',
      'someCpu',
      {
        cpu: {
          validMsg: 'CPUs must be a number',
          value: 'someCpu'
        }
      },
      false
    );
  });

  it('memory field validation is triggered', () => {
    testFormChange(
      'memory',
      'someMemory',
      {
        memory: {
          validMsg: 'Memory (GB) must be a number',
          value: 'someMemory'
        }
      },
      false
    );
  });

  it('is invalid when one required fields is missing', () => {
    const onChange = jest.fn();
    const component = shallow(testCreateVmWizard(validBasicSettings, onChange));
    component.instance().onFormChange('', 'namespace');

    expectMockToBeCalledWith(
      onChange,
      {
        ...validBasicSettings,
        namespace: {
          validMsg: 'Namespace is required',
          value: ''
        }
      },
      false
    );
  });
});
