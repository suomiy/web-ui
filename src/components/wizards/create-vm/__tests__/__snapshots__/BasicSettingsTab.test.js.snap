// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateVmWizard /> renders correctly 1`] = `
<Component
  fields={
    Object {
      "authKeys": Object {
        "isVisible": [Function],
        "required": true,
        "title": "Authenticated SSH Keys",
        "type": "textarea",
      },
      "cloudInit": Object {
        "title": "Use cloud-init",
        "type": "checkbox",
      },
      "cpu": Object {
        "isVisible": [Function],
        "required": true,
        "title": "CPUs",
        "validate": [Function],
      },
      "createTemplate": Object {
        "noBottom": true,
        "title": "Create new template from configuration",
        "type": "checkbox",
      },
      "description": Object {
        "title": "Description",
        "type": "textarea",
      },
      "flavor": Object {
        "choices": Array [
          Object {
            "name": "Flavor1",
          },
          Object {
            "name": "Flavor2",
          },
          Object {
            "name": "Custom",
          },
        ],
        "defaultValue": "--- Select Flavor ---",
        "id": "flavor-dropdown",
        "required": true,
        "title": "Flavor",
        "type": "dropdown",
      },
      "hostname": Object {
        "isVisible": [Function],
        "required": true,
        "title": "Hostname",
      },
      "imageSourceType": Object {
        "choices": Array [
          Object {
            "name": "PXE",
          },
          Object {
            "name": "URL",
          },
          Object {
            "name": "Template",
          },
          Object {
            "name": "Registry",
          },
        ],
        "defaultValue": "--- Select Provision Source ---",
        "id": "image-source-type-dropdown",
        "required": true,
        "title": "Provision Source",
        "type": "dropdown",
      },
      "imageURL": Object {
        "isVisible": [Function],
        "required": true,
        "title": "URL",
      },
      "memory": Object {
        "isVisible": [Function],
        "required": true,
        "title": "Memory (GB)",
        "validate": [Function],
      },
      "name": Object {
        "required": true,
        "title": "Name",
      },
      "namespace": Object {
        "choices": Array [
          Object {
            "name": "default",
          },
          Object {
            "name": "myproject",
          },
        ],
        "defaultValue": "--- Select Namespace ---",
        "id": "namespace-dropdown",
        "required": true,
        "title": "Namespace",
        "type": "dropdown",
      },
      "operatingSystem": Object {
        "choices": Array [
          Object {
            "id": "alpinelinux3.5",
            "name": "Alpine Linux 3.5",
          },
          Object {
            "id": "alpinelinux3.6",
            "name": "Alpine Linux 3.6",
          },
          Object {
            "id": "alpinelinux3.7",
            "name": "Alpine Linux 3.7",
          },
          Object {
            "id": "altlinux1.0",
            "name": "Mandrake RE Spring 2001",
          },
        ],
        "defaultValue": "--- Select Operating System ---",
        "id": "operating-system-dropdown",
        "required": true,
        "title": "Operating System",
        "type": "dropdown",
      },
      "registryImage": Object {
        "isVisible": [Function],
        "required": true,
        "title": "Registry Image",
      },
      "startVM": Object {
        "noBottom": true,
        "title": "Start virtual machine on creation",
        "type": "checkbox",
      },
      "template": Object {
        "choices": Array [
          Object {
            "name": "Template1",
          },
          Object {
            "name": "Template2",
          },
        ],
        "defaultValue": "--- Select Template ---",
        "id": "template-dropdown",
        "isVisible": [Function],
        "required": true,
        "title": "Template",
        "type": "dropdown",
      },
      "workloadProfile": Object {
        "choices": Array [
          Object {
            "description": "profile description",
            "id": "profile-one",
            "name": "WorkloadProfile1",
          },
          Object {
            "description": "profile description",
            "id": "profile-two",
            "name": "WorkloadProfile2",
          },
        ],
        "defaultValue": "--- Select Workload Profile ---",
        "help": [Function],
        "id": "workload-profile-dropdown",
        "required": true,
        "title": "Workload Profile",
        "type": "dropdown",
      },
    }
  }
  fieldsValues={Object {}}
  onFormChange={[Function]}
/>
`;
